### ISSUE TITLE
Ignoring the fact condition when the fact is not passed in

### ISSUE THREAD

**Issue by arsalans** [CacheControl/json-rules-engine#22]
Hi:

I would have liked the rule engine to ignore the fact condition if it is in "ANY" when the fact is not passed in, e.g.,

If my facts are these:

```
let facts = {
    gameDuration: 40
  };
```

And my rule is:

```
engine.addRule({
    conditions: {
      any: [{
        fact: 'personalFoulCount',
        operator: 'greaterThanInclusive',
        value: 5
      },
        {
          fact: 'gameDuration',
          operator: 'equal',
          value: 40
        }]
    },
    event: {  // define the event to fire when the conditions evaluate truthy
      type: 'fouledOut',
      params: {
        message: 'Player has fouled out!'
      }
    }
  });

```
I would have preferred the rule engine would not have errored out and ignore the fact that personalFoulCount is not there as it was in ANY condition clause.

**Comment by CacheControl** [comment]
thanks for the feedback @arsalans

Can you elaborate on why you'd prefer that approach?  i.e., what's desirable about it silently treating the undefined fact as falsey, as opposed to simply explicitly passing in a default value of `{personalFoulCount: 0}` to `run()`?

Getting into specifics on this particular instance, I think a default value makes a lot of sense, since any player on a team will _always_ have an integer value for their foul count, starting at 0.

**Comment by arsalans** [comment]
So we are working with a system which has 1000+ fields. It responds back
with only those fields which are filled at any given time. And we want to
validate only those fields that are present in the json at that time (say
50 random fields out of 1000 fields) . So we will write rules for all the
1000+ fields but we expect the rule engine to select the relevant rules for
those 50 fields and run only those rules.

Also, defaulting the fields and sending all of them is bad for performance.

Did I make my case?

Regards,
Arsalan Siddiqui, BE, MI
Cell : (416) 801-7646
Voip/Fax : (647) 367-4400
http://ca.linkedin.com/in/arsalan
http://www.i-architect.ca

On Jan 19, 2017 5:51 PM, "Cache Hamm" <notifications@github.com> wrote:

thanks for the feedback @arsalans <https://github.com/arsalans>

Can you elaborate on why you'd prefer that approach? i.e., what's desirable
about it silently treating the undefined fact as falsey, as opposed to
simply explicitly passing in a default value of {personalFoulCount: 0} to
run()?

Getting into specifics on this particular instance, I think a default value
makes a lot of sense, since any player on a team will *always* have an
integer value for their foul count, starting at 0.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/CacheControl/json-rules-engine/issues/22#issuecomment-273924307>,
or mute the thread

**Comment by CacheControl** [comment]
@arsalans yep, I think your situation makes the case for your request quite well.

We could certainly add an option to the engine constructor to turn off hard enforcement of fact existence, e.g.  ```let engine = new Engine(rules, { ignoreUnregisteredFacts: true })```

However, I want to be transparent about the capabilities of this project, and make sure it's the right solution for your task.  Although I'd love to someday make this project smart enough to optimally select the correct rules to run, that's not a current capability.  So if you had 1,000 rules and fed them into the rules engine, it would run all 1,000 (most would fail immediately on the first condition and therefore be very quick to execute).  I've done some load testing at this sort of volume and the results were fine for my particular use case at the time; if I recall the entire run completed in well under 10 milliseconds.  That may sound like a reasonable benchmark, but it's important to remember that Javascript blocks while doing CPU bound tasks.  I may eventually look into implementing a [web worker](https://github.com/audreyt/node-webworker-threads) model to offload the work from the main thread if there's enough interest, but that's currently not on the roadmap.  This project really is targeted as a lightweight solution for executing dozens or hundreds of rules (not thousands).  In any event, I'd encourage you to do some simple benchmarking before committing to this or any other rules engine library.

Now, depending on your situation, you may be able to do some basic filtering of your rules and selectively feed a subset into the engine to be run.  This pattern works well when there's an obvious property you can segment rules using.  For example, if you needed to run only the rules for the current customer, you could do something like:

```js
let rules = require('./all-my-rules') // entire subset
let targetCustomerId = req.params.customerId
let rulesToRun = rules.filter(rule => rule.customerId === targetCustomerId)
engine.run(rulesToRun)
```

Let me know your thoughts

**Comment by arsalans** [comment]
Thank you very much for your insights into this matter.

`ignoreUnregisteredFacts` flag will really come in handy.

We have been thinking about breaking down 1000 rules into multiple rulesets so that we are not running all the 1000 rules for every request. Your example for filtering helps, however, would you suggest a way where we can have lets say 20 rulesets and then 20 engines running for them (one engine for one ruleset)? Can we group rules by a ruleset name and then filter on the name of the ruleset instead of filtering on the field name?

Let me know your thoughts.

**Comment by CacheControl** [comment]
@arsalans The filtering you described makes sense to me and should work fine, provided you can identify which group(s) of rules to run (obviously if you have 10 batches of 100 rules each, if you have to run all 10 batches the performance will be the same as simply running all 1000 in a single engine).  I think as long as you're running hundreds of rules (not thousands, you'll be fine).

I can probably build in the new flag you require fairly easily.  Give me a couple days to clear some time in my schedule and I can probably get it built by sometime next week.

**Comment by arsalans** [comment]
@CacheControl Thank you very much, I am looking forward to test this out :)

You get a medal -> 🥇

**Comment by arsalans** [comment]
@CacheControl Just to confirm, what is going to be the outcome of the following when all the changes are made:

If my facts are these:

```
let facts = {
    gameDuration: 40
  };
```

And my rule is:

```
engine.addRule({
    conditions: {
      any: [{
        fact: 'personalFoulCount',
        operator: 'greaterThanInclusive',
        value: 5
      },
        {
          fact: 'gameDuration',
          operator: 'equal',
          value: 40
        }]
    },
    event: {  // define the event to fire when the conditions evaluate truthy
      type: 'fouledOut',
      params: {
        message: 'Player has fouled out!'
      }
    }
  });

```
I am expecting the following outcome (as the game duration fact is still valid):

`'Player has fouled out!'`

**Comment by arsalans** [comment]
In another scenario, if my facts are these:

```
let facts = {
    gameDuration: 40
  };
```

And my rule is:

```
engine.addRule({
    conditions: {
      all: [{
        fact: 'personalFoulCount',
        operator: 'greaterThanInclusive',
        value: 5
      },
        {
          fact: 'gameDuration',
          operator: 'equal',
          value: 40
        }]
    },
    event: {  // define the event to fire when the conditions evaluate truthy
      type: 'fouledOut',
      params: {
        message: 'Player has fouled out!'
      }
    }
  });

```
I am expecting that the rule will not produce an event as both facts are in the `ALL `condition and one is not present.

**Comment by CacheControl** [comment]
@arsalans yep, we're aligned.  Any unregistered fact will simply be treated as a falsey condition.  I'll probably attempt to come up with a better term than `ignoreUnregisteredFacts` that better explains the behavior

**Comment by CacheControl** [comment]
@arsalans this feature is now available in `1.4.0` as an `allowUndefinedFacts` engine option.  Please see the updated [docs](https://github.com/CacheControl/json-rules-engine/blob/master/docs/engine.md#options)
